File: node_runtime/src/kiro-session.js
import { spawn } from 'node:child_process';

export const SYSTEM_PROMPT = `You are a Tech Lead AI Agent.

You MUST:
- Manage tasks from Jira
- Communicate ONLY via Jira comments
- Implement code and create PR via Git MCP
- Continue work based on updates

Always track Task ID.`;

function wait(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export class KiroSession {
  constructor({ command, args = [], cwd, interactionDelayMs = 0 } = {}) {
    this.command = command;
    this.args = args;
    this.cwd = cwd;
    this.interactionDelayMs = interactionDelayMs;
    this.proc = null;
  }

  async start() {
    if (this.proc && !this.proc.killed) {
      return;
    }

    this.proc = spawn(this.command, this.args, {
      cwd: this.cwd,
      stdio: ['pipe', 'pipe', 'pipe'],
    });

    await this.send(SYSTEM_PROMPT);
  }

  async send(message) {
    if (!this.proc || this.proc.killed) await this.start();
    if (!this.proc?.stdin || this.proc.stdin.destroyed) {
      throw new Error('Kiro process is not available');
    }

    await new Promise((resolve, reject) => {
      this.proc.stdin.write(`${message}\n`, (error) => {
        if (error) reject(error);
        else resolve();
      });
    });

    if (this.interactionDelayMs > 0) {
      await wait(this.interactionDelayMs);
    }
  }

  stop() {
    if (this.proc && !this.proc.killed) {
      this.proc.kill('SIGTERM');
    }
  }
}

File: src/kiro_orchestration/kiro_session.py
from __future__ import annotations

import subprocess
from pathlib import Path

SYSTEM_PROMPT = """You are a Tech Lead AI Agent.

You MUST:
- Manage tasks from Jira
- Communicate ONLY via Jira comments
- Implement code and create PR via Git MCP
- Continue work based on updates

Always track Task ID.
"""


class KiroSession:
    def __init__(self, command: list[str], cwd: str | None = None) -> None:
        self.command = command
        self.cwd = cwd
        self.proc: subprocess.Popen[str] | None = None

    def start(self) -> None:
        if self.proc and self.proc.poll() is None:
            return

        self.proc = subprocess.Popen(
            self.command,
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            stderr=subprocess.STDOUT,
            text=True,
            cwd=Path(self.cwd) if self.cwd else None,
            bufsize=1,
        )
        self.send(SYSTEM_PROMPT)

    def send(self, message: str) -> None:
        if not self.proc or self.proc.poll() is not None:
            self.start()

        if not self.proc or not self.proc.stdin:
            raise RuntimeError("Kiro process is not available")

        self.proc.stdin.write(message + "\n")
        self.proc.stdin.flush()

    def stop(self) -> None:
        if self.proc and self.proc.poll() is None:
            self.proc.terminate()
            self.proc.wait(timeout=5)
